### PR TITLE
[cli] add pd state

### PR DESCRIPTION
**Pull request by sherysheng**
print out the DHCPv6 PD feature's state
enable/disable DHCPv6 PD

### CODE REVIEW

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -968,6 +971,50 @@ void Interpreter::OutputNat64Counters(const otNat64Counters &aCounters)
 
 #endif // OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE || OPENTHREAD_CONFIG_NAT64_BORDER_ROUTING_ENABLE
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_ENABLE
+template <> otError Interpreter::Process<Cmd("pd")>(Arg aArgs[])
```
Comment (erjiaqing) [inline src/cli/cli.cpp:975]: You should add this command as a subcommand of `br`

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -968,6 +971,50 @@ void Interpreter::OutputNat64Counters(const otNat64Counters &aCounters)
...
+     * Done
+     * @endcode
+     * @par
+     * Gets the state of DHCPv6 PD feature.
+     *
+     */
+    if (aArgs[0] == "state")
+    {
+        static const char *const kDhcpv6PdState[] = {"Disabled", "Stopped", "Running"};
+        static_assert(0 == OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED, "DHCPv6 PD is disabled on the border router");
```
Comment (erjiaqing) [inline src/cli/cli.cpp:1000]:
```suggestion
        static_assert(0 == OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED, "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED is not expected!");
```

Similar for following static_asserts.

File: src/cli/cli.cpp [hunk 3]
```diff
@@ -968,6 +971,50 @@ void Interpreter::OutputNat64Counters(const otNat64Counters &aCounters)
...
+     * @cli pd state
+     * @code
+     * pd state
+     * Disabled
+     * Stopped
+     * Running
+     * Done
+     * @endcode
+     * @par
+     * Gets the state of DHCPv6 PD feature.
```
Comment (erjiaqing) [inline src/cli/cli.cpp:994]:
```suggestion
    /**
     * @cli pd state
     * @code
     * pd state
     * Running
     * Done
     * @endcode
     * @par api_copy
     * #otBorderRoutingDhcp6PdGetState
     */
```

File: src/cli/README.md [hunk 4]
```diff
@@ -2791,6 +2792,16 @@ Set the preferred Thread Leader Partition ID.
 Done
 ```
 
+### pd state
+
+Get the state of DHCPv6 PD feature.
```
Comment (erjiaqing) [inline src/cli/README.md:2797]:
```suggestion
Get the state of DHCPv6 PD.
```

Or PD prefix manager, or just PD.

**Review by superwhd — APPROVED** [review]
LGTM 👍

File: src/cli/cli_br.cpp [hunk 5]
```diff
@@ -435,6 +435,45 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+     * br pd state
+     * running
+     * Done
+     * @endcode
+     * @par api_copy
+     * #otBorderRoutingDhcp6PdGetState
+     */
+    if (aArgs[0] == "state")
+    {
+        static const char *const kDhcpv6PdState[] = {"disabled", "stopped", "running"};
```
Comment (abtink) [inline src/cli/cli_br.cpp:457]: To follow similar style (similar to `Br::Process<Cmd("state")>(Arg aArgs[])`):
```suggestion
        static const char *const kDhcpv6PdStateStrings[] =
        {
            "disabled", // (0) OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED
            "stopped",  // (1) OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED
            "running",  // (2) OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING
        };

```

File: src/cli/cli_br.cpp [hunk 6]
```diff
@@ -435,6 +435,45 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+        static_assert(0 == OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED,
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED is not expected!");
+        static_assert(1 == OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED,
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED is not expected!");
+        static_assert(2 == OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING,
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING is not expected!");
```
Comment (abtink) [inline src/cli/cli_br.cpp:463]: Shorter lines to keep it a single line?
```suggestion
        static_assert(0 == OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED, "DHCP6_PD_STATE_DISABLED value is incorrect");
        static_assert(1 == OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED, "DHCP6_PD_STATE_STOPPED value is incorrect");
        static_assert(2 == OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING, "DHCP6_PD_STATE_RUNNING value is incorrect);
```

File: src/cli/cli_br.cpp [hunk 7]
```diff
@@ -435,6 +435,45 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+     * running
+     * Done
+     * @endcode
+     * @par api_copy
+     * #otBorderRoutingDhcp6PdGetState
+     */
+    if (aArgs[0] == "state")
+    {
+        static const char *const kDhcpv6PdState[] = {"disabled", "stopped", "running"};
+        static_assert(0 == OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED,
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/cli/cli_br.cpp:458]: We are relying on enum values to do string look up using array index.
This is to ensure if the enum value get changed (new state added, etc), it causes a build error and does not go unnoticed.

File: src/cli/cli_br.cpp [hunk 8]
```diff
@@ -435,6 +435,45 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED is not expected!");
+        static_assert(1 == OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED,
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED is not expected!");
+        static_assert(2 == OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING,
+                      "The value of OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING is not expected!");
+
+        OutputLine("%s", Stringify(otBorderRoutingDhcp6PdGetState(GetInstancePtr()), kDhcpv6PdState));
+    }
+    else
```
Comment (abtink) [inline src/cli/cli_br.cpp:467]: Should we add "br pd enable/disable" as well?

**Review by abtink — COMMENTED** [review]
Thanks. 
Some suggestions below:

File: src/cli/cli_br.cpp (hunk 8 above)
Comment (sherysheng) [reply src/cli/cli_br.cpp:467]: we don't have API yet, plan to add enable/disable next time

File: src/cli/cli_br.cpp [hunk 9]
```diff
@@ -435,6 +435,45 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
     return error;
 }
 
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_DHCP6_PD_ENABLE
+template <> otError Br::Process<Cmd("pd")>(Arg aArgs[])
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aArgs[0].IsEmpty(), error = OT_ERROR_INVALID_ARGS);
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply src/cli/cli_br.cpp:443]: aha yes!! thanks!!

File: src/cli/cli_br.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/cli/cli_br.cpp:458]: Got it. Thanks!

**Review by Irving-cl — APPROVED** [review]
LGTM 👍

File: src/cli/cli_br.cpp (hunk 8 above)
Comment (abtink) [reply src/cli/cli_br.cpp:467]: We have `otBorderRoutingDhcp6PdSetEnabled()`?

Comment (sherysheng) [reply src/cli/cli_br.cpp:467]: ohh cool, I will add it later!

**Review by abtink — COMMENTED** [review]
Looks good. Thanks 👍 
One small issue above with missing `else`.

File: src/cli/cli_br.cpp [hunk 10]
```diff
@@ -435,6 +435,66 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+     * @cli br pd state
+     * @code
+     * br pd state
+     * running
+     * Done
+     * @endcode
+     * @par api_copy
+     * #otBorderRoutingDhcp6PdGetState
+     */
+    if (aArgs[0] == "state")
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply src/cli/cli_br.cpp]: thanks！！

**Review by erjiaqing — APPROVED** [review]

File: src/cli/cli_br.cpp [hunk 11]
```diff
@@ -435,6 +435,66 @@ template <> otError Br::Process<Cmd("prefixtable")>(Arg aArgs[])
...
+     * br pd disable
+     * Done
+     * @endcode
+     * @cparam br pd @ca{enable|disable}
+     * @par api_copy
+     * #otBorderRoutingDhcp6PdSetEnabled
+     *
+     */
+    if (Interpreter::GetInterpreter().ProcessEnableDisable(aArgs, otBorderRoutingDhcp6PdSetEnabled) == OT_ERROR_NONE)
+    {
```
Comment (erjiaqing) [inline src/cli/cli_br.cpp:459]: We already have `otBorderRoutingDhcp6PdSetEnabled`, so we can call it now?

Comment (erjiaqing) [reply src/cli/cli_br.cpp:459]: I did not see a call of `otBorderRoutingDhcp6PdSetEnabled(GetInstancePtr(), otBorderRoutingDhcp6PdSetEnabled)` here?

Am I missing anything?

Comment (erjiaqing) [reply src/cli/cli_br.cpp:459]: Ah, I misunderstood the function, sorry.

**Review by abtink — APPROVED** [review]
LGTM

**Review by Vyrastas — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**